ua-snap/data-api
# Add service endpoints for the most recent GIPL 2.0 outputs

This PR adds service endpoints for the most recent of GIPL model outputs (ten variables at 1 km spatial resolution) that were provided to SNAP by Sergey Marchenko in August 2022. The backend data supporting these endpoints is a Rasdaman coverage called `crrel_gipl_outputs` (currently on Apollo only) that was created by ingesting 6000 GeoTIFFs. The spatial extent of these data are Alaska only - although the Western Aleutian chain is not present. 

There are 3 model cases (5-Model-Average, GFDL-CM3, and NCAR-CCSM4), and two scenarios (RCPs 4.5 and 8.5).

Only point queries are implemented at this time.

To test this PR, try queries constructed as follows to fetch the complete set of GIPL model outputs (all models, scenarios, and years) for a point location:

http://localhost:5000/permafrost/point/gipl/66.0628/-146.1627

To slice a time span from these annual-frequency data, simply append a start and stop year to your path - try these examples to fetch data and test validation:

http://localhost:5000/permafrost/point/gipl/65/-148/2040/2049
http://localhost:5000/permafrost/point/gipl/65/-148/2049/2040
http://localhost:5000/permafrost/point/gipl/65/-148/2040/3049
http://localhost:5000/permafrost/point/gipl/65/-148/1040/2049

To slice a time span from these annual-frequency data and then compute mean-min-max summaries for each of the ten variables, simply append "mmm" to your URL:

http://localhost:5000/permafrost/point/gipl/65/-148/2040/2069/mmm

To test validation:

http://localhost:5000/permafrost/point/gipl/65/-148/2040/2069/foo

Highlights:
 - Check the new WCPS syntax courtesy of Dimitar! Pretty clean, and easy to flip between min / mean / max.
 - I like how this route came together: all data > time slice > time slice summarized

Notes:
 - I tread carefully to make sure the other permafrost data (including the Melvin et al 4 km GIPL outputs served used in NCR) endpoints were not impacted. But keep an eye out for that.
 - Also, take a close look at the template. Are the GIPL outputs distinct enough? I think we'd love to mothball the 4 km stuff, but that could be work because of the NCR.
 - In the code I've tried to distinguish the new GIPL data by using `gipl1km` everywhere.

Misses:
 - We need to add to the Data Source table. You have to scroll far to even get to that table, but it is important. This means adding the 6000 GeoTIFFs to our GeoNetwork Catalog.
